feat: Add Read Support for Presto-MaxCompute Connector

dingxin-tech opened the pull request:
This PR introduces the read support for the Presto-MaxCompute Connector, enabling users to perform queries on tables within Alibaba's MaxCompute.
While the current state of the connector does not support write operations, DDL operations, or querying partitioned tables, it is a significant step forward in integrating MaxCompute with Presto.

Key features included in this PR:
- Read support for MaxCompute standard tables.
- Configuration setup to connect Presto with MaxCompute projects.
- Documentation updates for setting up and using the connector.

The following are yet to be implemented in future PRs:
- The ability to query partitioned tables.
- Write support for MaxCompute tables.
- Support for DDL operations within MaxCompute from Presto.

dingxin-tech committed:
init maxcompute-presto

dingxin-tech committed:
init maxcompute-presto

dingxin-tech committed:
feat: support array type and add doc

dingxin-tech committed:
doc: fix doc

dingxin-tech committed:
style: update gitignore

dingxin-tech committed:
fix: result error and add some logs

tdcmeehan commented:
@dingxin-tech thank you for your contribution.  Regarding testing: can you describe how to add tests for this plugin?  Is it possible to add integration tests that run against a mocked version of MaxCompute, or through some other means?

dingxin-tech commented:
@tdcmeehan I have no idea how to add integration tests since MaxCompute currently does not provide external testing accounts. Do you know how other similar databases handle this? Maybe I should implement something like a MaxCompute emulator similar to the [BigQuery emulator](https://github.com/goccy/bigquery-emulator)?

tdcmeehan commented:
@dingxin-tech ideally, Alibaba would give us a testing token to just run some minimal workload.  If you could create an emulator, that would be awesome.

dingxin-tech commented:
> @dingxin-tech ideally, Alibaba would give us a testing token to just run some minimal workload. If you could create an emulator, that would be awesome.

Yes, I am currently working on the development of the MaxCompute-Emulator, and I am pleased to report that we have made some significant incremental progress. With the current pace, I anticipate that we will be able to complete support for Presto regression within the next two to three weeks.

dingxin-tech committed:
fix: upper case column name result error

dingxin-tech committed:
test: add maxcompute-emulator and some e2e test

tdcmeehan commented:
@dingxin-tech I see you've added an emulator, is this ready to review?

dingxin-tech commented:
> @dingxin-tech I see you've added an emulator, is this ready to review?

Thank you for your attention. 

Yes, I believe this PR is ready for review.

Although some critical features (such as reading the partition table, writing to the table) are not supported yet (in fact, the emulator also does not support these functionalities at the moment), I believe the current functionalities form a good initial whole for the first commit in this repository. These new features can be added in subsequent PRs to further improve the connector.

dingxin-tech committed:
update version to 0.287

dingxin-tech committed:
feat: support read partitioned table

tdcmeehan commented on src/main/java/com/facebook/presto/maxcompute/MaxComputeInputSplit.java:
```diff
@@ -0,0 +1,103 @@
+    private Long numRecord;
+
+    @JsonCreator
+    public MaxComputeInputSplit(@JsonProperty("sessionId") String sessionId, @JsonProperty("splitIndex") Integer splitIndex,
```
```suggestion
    public MaxComputeInputSplit(
            @JsonProperty("sessionId") String sessionId, 
            @JsonProperty("splitIndex") Integer splitIndex,
```

tdcmeehan commented on src/main/resources/META-INF/services/com.facebook.presto.spi.Plugin:
```diff
@@ -0,0 +1 @@
+com.facebook.presto.maxcompute.MaxComputePlugin
```
Can you use the presto-maven-plugin to add this so you don't need to add it by hand?

dingxin-tech replied:
OK, I am testing the integration of my Runner into `AbstractTestIntegrationSmokeTest`. I found that `AbstractTestIntegrationSmokeTest` requires a TPCH dataset in my database emulator, so I generated one with TPCH datagen. However, I found that the generated dataset does not match the test expectations. For example, in the `testAggregateSingleColumn` test, there is a java.lang.AssertionError:

```
For query:
 SELECT SUM(orderkey) FROM orders
not equal
Actual rows (1 of 1 extra rows shown, 1 rows in total):
    [4499987250000]
Expected rows (1 of 1 missing rows shown, 1 rows in total):
    [449872500]
```

Can you tell me how the TPCH dataset required by this test is generated? Does it use a sample ratio of 0.001?

tdcmeehan replied:
[We use the tiny scale factor, which is 0.01.](https://github.com/prestodb/presto/blob/8f26ac6548ff738940ffa4e9b34fed209878a22e/presto-tpch/src/main/java/com/facebook/presto/tpch/TpchMetadata.java#L96-L97)

tdcmeehan reviewed (changes requested):
Really great work.  Can you please add integration style tests and more unit tests for the various utilities?  For example, I think we should have decent unit test coverage over the Arrow to Page conversions.  I also think we can add a dependency on `presto-tests`, and run `AbstractTestIntegrationSmokeTest` using your new query runner.  Can you give it a try?

dingxin-tech committed:
style: apply suggested change

Co-authored-by: Timothy Meehan <tim@timdmeehan.com>

dingxin-tech committed:
feat(maxcompute): Update the MaxCompute simulator image and add integrated smoke testing

Update dependencies to support the new version of the MaxCompute simulator image, and introduce new test classes to verify basic integration behavior. At the same time, Maven configuration is enhanced to better support the Presto plug-in development process.

dingxin-tech commented:
I have made changes based on your suggestions: 

1. I removed the configuration from META-INF and added a dependency on the `presto-maven-plugin`. 
2. I upgraded the emulator and added `TestIntegrationSmokeTest`. 
3. Regarding the additional tests for ArrowToPage that you mentioned, I currently have no idea on how to determine if a Page meets expectations, but I believe the SmokeTest has already covered that part of the cases. 

Could you please review it again? @tdcmeehan

tdcmeehan commented:
Can you please add a Github action to trigger the tests to run in CI?  See here for reference: https://github.com/prestodb/airlift/pull/77/files

elharo commented on pom.xml:
```diff
@@ -0,0 +1,240 @@
+        <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+            <version>26.0-jre</version>
```
can push this to the current version?

dingxin-tech replied:
sure.

elharo commented on src/checkstyle/presto-checks.xml:
```diff
@@ -0,0 +1,222 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
I'm not sure this file belongs in this PR

dingxin-tech replied:
maybe we should add it ?

elharo commented on src/main/java/com/facebook/presto/maxcompute/utils/ArrowUtils.java:
```diff
@@ -0,0 +1,314 @@
+                }
+                return new SimpleStruct((StructTypeInfo) structTypeInfo, values);
+            }
+            catch (Exception e) {
```
can you be more specific about the exception?

dingxin-tech replied:
Sure,  I realized that this method doesn't actually throw a checked exception, so I've removed the try-catch block.  I'm not sure why I implemented it that way :)

elharo commented on src/main/java/com/facebook/presto/maxcompute/utils/ArrowUtils.java:
```diff
@@ -0,0 +1,314 @@
+                return new SimpleStruct((StructTypeInfo) structTypeInfo, values);
+            }
+            catch (Exception e) {
+                throw new RuntimeException("Could not get the struct", e);
```
don't throw raw runtime exceptions, use a subclass

elharo commented on src/main/java/com/facebook/presto/maxcompute/utils/CommonUtils.java:
```diff
@@ -0,0 +1,87 @@
+
+    private CommonUtils() {}
+
+    public static String serialize(Serializable object)
```
why is object serialization used at all? It's mostly avoided in java these days as it's been a common source of security issues

dingxin-tech replied:
The `serialize` and `deserialize` methods are specifically utilized for serializing and deserializing `TableBatchReadSession` objects. This is particularly important for `MaxComputeSplit`, where JSON format is employed for message transmission within Presto. Given that `TableBatchReadSession` contains numerous variables and has been well-implemented with the `Serializable` interface, I believe utilizing these methods is a sound approach.

elharo commented on src/main/java/com/facebook/presto/maxcompute/utils/CommonUtils.java:
```diff
@@ -0,0 +1,87 @@
+                    throw new IOException("Failed after retries", e);
+                }
+                try {
+                    LOG.warn("Retrying after exception: " + e.getMessage(), e);
```
Is this actionable? Will anyone read it?

dingxin-tech replied:
The method was actually not being used, so I've removed it.

elharo reviewed (commented):
What is our current thinking about adding new connectors to the core repo as opposed to third party repos? Is there anything preventing this from being a separate project?

tdcmeehan commented:
> What is our current thinking about adding new connectors to the core repo as opposed to third party repos? Is there anything preventing this from being a separate project?

There's nothing preventing this from being a separate project, however we would like to move more of our modules in `presto` to third party modules, and given that MaxCompute is a widely used and publicly available offering, I'm OK adding this to the Presto org.  I also think it will be a good model for other connectors to follow.

steveburnett commented on README.md:
```diff
@@ -1 +1,96 @@
-# presto-alibaba-maxcompute-connector
+# MaxCompute Connector
+
+The MaxCompute connector enables direct querying and processing of data stored in the MaxCompute data warehouse. This connector is particularly useful for implementing data
```
```suggestion
The MaxCompute connector enables direct querying and processing of data stored in the MaxCompute data warehouse. This connector is useful for implementing data
```

steveburnett commented on README.md:
```diff
@@ -1 +1,96 @@
+
+### Prerequisites
+
+Before using the MaxCompute connector, please ensure the following conditions are met:
```
```suggestion
Before using the MaxCompute connector, ensure the following conditions are met:
```

steveburnett commented on README.md:
```diff
@@ -1 +1,96 @@
+
+Before using the MaxCompute connector, please ensure the following conditions are met:
+
+- Access permissions: Ensure you have valid access rights to a MaxCompute project and that you have obtained the necessary authentication credentials.
```
```suggestion
- Access permissions: You must have valid access rights to a MaxCompute project and the necessary authentication credentials for that project.
```

steveburnett commented on README.md:
```diff
@@ -1 +1,96 @@
+Before using the MaxCompute connector, please ensure the following conditions are met:
+
+- Access permissions: Ensure you have valid access rights to a MaxCompute project and that you have obtained the necessary authentication credentials.
+- Network setup: For stable data transmission, the use of Aliyun VPC network is recommended. Check out
```
```suggestion
- Network setup: For stable data transmission, the use of Aliyun VPC network is recommended. See
```

steveburnett commented on README.md:
```diff
@@ -1 +1,96 @@
+- Access permissions: Ensure you have valid access rights to a MaxCompute project and that you have obtained the necessary authentication credentials.
+- Network setup: For stable data transmission, the use of Aliyun VPC network is recommended. Check out
+  the [MaxCompute Network Configuration Guide](https://help.aliyun.com/zh/maxcompute/user-guide/network-connection-process) for more information.
+- Resource group configuration: Possession of an exclusive Tunnel resource group is required (although MaxCompute currently allows access to shared resource groups, this could
```
```suggestion
- Resource group configuration: Possession of an exclusive Tunnel resource group is required. Note: MaxCompute currently allows access to shared resource groups, but this could
```

steveburnett commented on README.md:
```diff
@@ -1 +1,96 @@
+- Network setup: For stable data transmission, the use of Aliyun VPC network is recommended. Check out
+  the [MaxCompute Network Configuration Guide](https://help.aliyun.com/zh/maxcompute/user-guide/network-connection-process) for more information.
+- Resource group configuration: Possession of an exclusive Tunnel resource group is required (although MaxCompute currently allows access to shared resource groups, this could
+  change at any time). [Exclusive resource group](https://help.aliyun.com/zh/maxcompute/user-guide/purchase-and-use-exclusive-resource-groups-for-dts).
```
```suggestion
  change at any time. See [Exclusive resource group](https://help.aliyun.com/zh/maxcompute/user-guide/purchase-and-use-exclusive-resource-groups-for-dts).
```

steveburnett commented on README.md:
```diff
@@ -1 +1,96 @@
+  the [MaxCompute Network Configuration Guide](https://help.aliyun.com/zh/maxcompute/user-guide/network-connection-process) for more information.
+- Resource group configuration: Possession of an exclusive Tunnel resource group is required (although MaxCompute currently allows access to shared resource groups, this could
+  change at any time). [Exclusive resource group](https://help.aliyun.com/zh/maxcompute/user-guide/purchase-and-use-exclusive-resource-groups-for-dts).
+- (Optional) Schema activation: Enable Schema mode to use different directories in Presto. Refer
```
```suggestion
- (Optional) Schema activation: Enable Schema mode to use different directories in Presto. See
```

steveburnett commented on README.md:
```diff
@@ -1 +1,96 @@
+- Resource group configuration: Possession of an exclusive Tunnel resource group is required (although MaxCompute currently allows access to shared resource groups, this could
+  change at any time). [Exclusive resource group](https://help.aliyun.com/zh/maxcompute/user-guide/purchase-and-use-exclusive-resource-groups-for-dts).
+- (Optional) Schema activation: Enable Schema mode to use different directories in Presto. Refer
+  to [How to Enable Schema Model in MaxCompute](https://help.aliyun.com/zh/maxcompute/user-guide/schema-related-operations) for more information.
```
```suggestion
  [How to Enable Schema Model in MaxCompute](https://help.aliyun.com/zh/maxcompute/user-guide/schema-related-operations) for more information.
```

steveburnett commented on README.md:
```diff
@@ -1 +1,96 @@
+
+### Obtain the Connector JAR
+
+To use the MaxCompute connector, you first need to obtain the JAR file by building the repository. Use the Maven command `mvn clean package` to build the project and generate the
```
```suggestion
Obtain the JAR file by building the repository. Use the Maven command `mvn clean package` to build the project and generate the
```
Consider making this a set of numbered steps. Something like the following example: 

1. Clone the repository (include a link if possible) 
2. Use the Maven command `mvn clean package` to build the project and generate the JAR. 
3. Move the JAR file into the `plugin/maxcompute` directory in your Presto installation.

steveburnett commented on README.md:
```diff
@@ -1 +1,96 @@
+
+### Configuration
+
+Before configuring the MaxCompute connector, please ensure you have prepared all the necessary information associated with your MaxCompute project, including project name, access
```
```suggestion
Before configuring the MaxCompute connector, have ready all the necessary information associated with your MaxCompute project, including project name, access
```

steveburnett commented on README.md:
```diff
@@ -1 +1,96 @@
+### Configuration
+
+Before configuring the MaxCompute connector, please ensure you have prepared all the necessary information associated with your MaxCompute project, including project name, access
+ID, access key, and resource group name.
```
```suggestion
ID, access key, and resource group name.

```
Suggest a blank line here to left-justify the next sentence "Follow the steps below..."

steveburnett commented on README.md:
```diff
@@ -1 +1,96 @@
+
+Before configuring the MaxCompute connector, please ensure you have prepared all the necessary information associated with your MaxCompute project, including project name, access
+ID, access key, and resource group name.
+Follow the steps below to create a configuration file (e.g., `maxcompute.properties`) in Presto's `etc/catalog` directory:
```
```suggestion
1. Create a configuration file (for example, `maxcompute.properties`) in Presto's `etc/catalog` directory:
```
Suggest making this a set of numbered steps.

steveburnett commented on README.md:
````diff
@@ -1 +1,96 @@
+odps.quota.name=<Your MaxCompute Exclusive Resource Group Name>
+```
+
+Replace `<Your MaxCompute Project Name>`, `<Your Alibaba Cloud Access ID>`, `<Your Alibaba Cloud Access Key>`, `<Your MaxCompute Service Endpoint>`,
````
```suggestion
2. Replace `<Your MaxCompute Project Name>`, `<Your Alibaba Cloud Access ID>`, `<Your Alibaba Cloud Access Key>`, `<Your MaxCompute Service Endpoint>`,
```

steveburnett commented on README.md:
````diff
@@ -1 +1,96 @@
+```
+
+Replace `<Your MaxCompute Project Name>`, `<Your Alibaba Cloud Access ID>`, `<Your Alibaba Cloud Access Key>`, `<Your MaxCompute Service Endpoint>`,
+and `<Your MaxCompute Exclusive Resource Group Name>` with the actual values for your MaxCompute environment.
````
```suggestion
and `<Your MaxCompute Exclusive Resource Group Name>` with the values for your MaxCompute environment.
```

steveburnett commented on README.md:
```diff
@@ -1 +1,96 @@
+
+### Accessing Multiple MaxCompute Projects
+
+To access multiple MaxCompute projects simultaneously, please create separate configuration files for each project. For example, if you have two projects, one for "sales" (`sales`)
```
```suggestion
To access multiple MaxCompute projects simultaneously, create separate configuration files for each project. For example, if you have two projects, one for "sales" (`sales`)
```

steveburnett commented on README.md:
```diff
@@ -1 +1,96 @@
+
+- `sales.properties` for the sales project.
+- `analytics.properties` for the analysis project.
+  With this setup, `sales` and `analytics` directories will be created in Presto respectively.
```
```suggestion
  With this setup, `sales` and `analytics` directories will be created in Presto respectively.
```
Should `directories` be `catalogs`?

steveburnett commented on README.md:
```diff
@@ -1 +1,96 @@
+
+### Data Type Mapping
+
+When using the MaxCompute connector, MaxCompute's data types will be mapped to Presto's data types. The table below shows this mapping relationship and its corresponding Presto
```
```suggestion
When using the MaxCompute connector, MaxCompute's data types are mapped to Presto's data types. The table below shows this mapping relationship and its corresponding Presto
```

steveburnett commented on README.md:
```diff
@@ -1 +1,96 @@
+- `analytics.properties` for the analysis project.
+  With this setup, `sales` and `analytics` directories will be created in Presto respectively.
+
+### Data Type Mapping
```
This Data Type Mapping section would be ideal to move into the connector documentation for this new connector, in a new file in https://github.com/prestodb/presto/tree/master/presto-docs/src/main/sphinx/connector.

steveburnett reviewed (changes requested):
Thanks for the documentation! I commented with several nits about phrasing, and a couple of suggestions about formatting. If any of my suggestions don't work for you, I'm happy to discuss them if you like and explain my reasoning. 

Also, I suggest you add a new page to the Presto documentation for the new connector, by creating a new file in https://github.com/prestodb/presto/tree/master/presto-docs/src/main/sphinx/connector and adding that new file to the [Connectors](https://github.com/prestodb/presto/blob/master/presto-docs/src/main/sphinx/connector.rst) index file.

dingxin-tech committed:
chore: Add Maven build process configuration

dingxin-tech committed:
chore(pom.xml): update dependency version number

- Upgraded `odps.sdk.version` from 0.48.5-public to 0.49.0-public
- Upgrade guava version from 26.0-jre to 33.3.1-jre

dingxin-tech committed:
refactor(ArrowUtils): remove unused RuntimeException

dingxin-tech committed:
refactor(CommonUtils): remove unused function

dingxin-tech committed:
Update README.md

Co-authored-by: Steve Burnett <burnett@pobox.com>

dingxin-tech committed:
Update README.md

Co-authored-by: Steve Burnett <burnett@pobox.com>

dingxin-tech committed:
Update README.md

Co-authored-by: Steve Burnett <burnett@pobox.com>

dingxin-tech committed:
Update README.md

Co-authored-by: Steve Burnett <burnett@pobox.com>

dingxin-tech committed:
Update README.md

Co-authored-by: Steve Burnett <burnett@pobox.com>

dingxin-tech committed:
Update README.md

Co-authored-by: Steve Burnett <burnett@pobox.com>

dingxin-tech committed:
Update README.md

Co-authored-by: Steve Burnett <burnett@pobox.com>

dingxin-tech committed:
Update README.md

Co-authored-by: Steve Burnett <burnett@pobox.com>

dingxin-tech committed:
update readme.md

dingxin-tech commented:
> Thanks for the documentation! I commented with several nits about phrasing, and a couple of suggestions about formatting. If any of my suggestions don't work for you, I'm happy to discuss them if you like and explain my reasoning.
> 
> Also, I suggest you add a new page to the Presto documentation for the new connector, by creating a new file in https://github.com/prestodb/presto/tree/master/presto-docs/src/main/sphinx/connector and adding that new file to the [Connectors](https://github.com/prestodb/presto/blob/master/presto-docs/src/main/sphinx/connector.rst) index file.

Thank you for your input; all of your suggestions have been incorporated into the changes. I agree that it would be beneficial to add a link to the MaxCompute connector in the Presto documentation repository. However, this should be done in a separate PR, potentially initiated after this current PR has been merged.

dingxin-tech committed:
update test to adapt new sdk version

dingxin-tech commented:
> Really great work. Can you please add integration style tests and more unit tests for the various utilities? For example, I think we should have decent unit test coverage over the Arrow to Page conversions. I also think we can add a dependency on `presto-tests`, and run `AbstractTestIntegrationSmokeTest` using your new query runner. Can you give it a try?

Sure, I've added. :)

steveburnett commented:
> Thank you for your input; all of your suggestions have been incorporated into the changes. I agree that it would be beneficial to add a link to the MaxCompute connector in the Presto documentation repository. However, this should be done in a separate PR, potentially initiated after this current PR has been merged.

Thank you! Some of my comments may have been hidden by GitHub. When you have time, would you please recheck? They may be hidden under "8 hidden conversations". 
<img width="858" alt="Screenshot 2024-09-27 at 11 15 31 AM" src="https://github.com/user-attachments/assets/e8c3c952-0a57-4040-b8c3-11449722186d">

dingxin-tech committed:
Update README.md

Co-authored-by: Steve Burnett <burnett@pobox.com>

dingxin-tech committed:
Update README.md

Co-authored-by: Steve Burnett <burnett@pobox.com>

dingxin-tech committed:
Update README.md

Co-authored-by: Steve Burnett <burnett@pobox.com>

dingxin-tech committed:
Update README.md

Co-authored-by: Steve Burnett <burnett@pobox.com>

dingxin-tech committed:
Update README.md

Co-authored-by: Steve Burnett <burnett@pobox.com>

dingxin-tech committed:
Update README.md

Co-authored-by: Steve Burnett <burnett@pobox.com>

dingxin-tech committed:
Update README.md

Co-authored-by: Steve Burnett <burnett@pobox.com>

dingxin-tech committed:
Update README.md

Co-authored-by: Steve Burnett <burnett@pobox.com>

dingxin-tech commented:
> > Thank you for your input; all of your suggestions have been incorporated into the changes. I agree that it would be beneficial to add a link to the MaxCompute connector in the Presto documentation repository. However, this should be done in a separate PR, potentially initiated after this current PR has been merged.
> 
> Thank you! Some of my comments may have been hidden by GitHub. When you have time, would you please recheck? They may be hidden under "8 hidden conversations". <img alt="Screenshot 2024-09-27 at 11 15 31 AM" width="858" src="https://private-user-images.githubusercontent.com/7013443/371587869-e8c3c952-0a57-4040-b8c3-11449722186d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAxODg4NjEsIm5iZiI6MTczMDE4ODU2MSwicGF0aCI6Ii83MDEzNDQzLzM3MTU4Nzg2OS1lOGMzYzk1Mi0wYTU3LTQwNDAtYjhjMy0xMTQ0OTcyMjE4NmQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMjlUMDc1NjAxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjcyMjQ0NGQ2Mjc2MzEyNDQwNDU4YzNhMzY2NjIzZjc4ODM5NTMwZDAwNWM3YWMwNjhkNDE1ZGE5ZjE1Y2JkZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Ldud4SeTgV_Bl_q-N-mmkXxmdW13sDB77zE_ucPxTRo">

Done!

steveburnett commented on README.md:
```diff
@@ -1 +1,97 @@
+file:
+
+- `sales.properties` for the sales project.
+- `analytics.properties` for the analysis project.
```
```suggestion
- `analytics.properties` for the analysis project.

```
Suggest add a blank line to left-justify the next line ("With this setup, ..."). As it is, "With this setup," is part of the `analytics.properties` list item.

steveburnett commented on README.md:
````diff
@@ -1 +1,97 @@
+
+1. Create a configuration file (for example, `maxcompute.properties`) in Presto's `etc/catalog` directory:
+
+```properties
````
Suggest indenting the code block so that it displays within item 1. As it is, the code block is flush with the left margin. You can check this with "View file" in GitHub, which is in the menu displayed when "..." in the upper right of this file is selected.

steveburnett commented on README.md:
```diff
@@ -1 +1,97 @@
+| DECIMAL              | DECIMAL          |
+| ARRAY                | ARRAY            |
+
+Please note that some MaxCompute data types (such as `MAP`, `STRUCT`) are not yet fully supported in the current version of the connector. We plan to continue refining support for
```
```suggestion
Note: Some MaxCompute data types (such as `MAP`, `STRUCT`) are not fully supported in the current version of the connector.
```
In general, documentation should describe the current state and does not use future tense. Suggest deleting the second sentence.

steveburnett commented on README.md:
```diff
@@ -1 +1,97 @@
+| ARRAY                | ARRAY            |
+
+Please note that some MaxCompute data types (such as `MAP`, `STRUCT`) are not yet fully supported in the current version of the connector. We plan to continue refining support for
+these data types in future versions.
```
```suggestion
```

steveburnett commented on README.md:
```diff
@@ -1 +1,97 @@
+
+### Notices
+
+This project is still under active development, with plans to introduce more features in subsequent versions, such as support for DDL operations, data writing, and expanded data
```
```suggestion
This project is under active development, with plans to introduce more features in subsequent versions, such as support for DDL operations, data writing, and expanded data
```
Suggest removing "still" as unneeded.

steveburnett reviewed (changes requested):
Thanks for the doc update! A few more suggestions, nothing major.

dingxin-tech committed:
Update README.md

Co-authored-by: Steve Burnett <burnett@pobox.com>

dingxin-tech committed:
Update README.md

Co-authored-by: Steve Burnett <burnett@pobox.com>

dingxin-tech committed:
Update README.md

Co-authored-by: Steve Burnett <burnett@pobox.com>

dingxin-tech committed:
Update README.md

Co-authored-by: Steve Burnett <burnett@pobox.com>

dingxin-tech committed:
Update README.md

dingxin-tech commented:
> Thanks for the doc update! A few more suggestions, nothing major.

Done.

steveburnett reviewed (approved):
LGTM! (docs) 

View file in GitHub to review it as it will be formatted when merged, everything looks good. Thanks!
